src/backends/perf_pt/perf_pt.c

```diff
@@ -348,124 +318,173 @@ tracer_thread(void *arg)
  */
 
 /*
- * Turn on Intel PT.
- *
- * Arguments:
- *   tr_conf: Tracer configuration.
- *
- * Returns a pointer to a tracer context, or NULL on error.
+ * Initialise a tracer context.
  */
 struct tracer_ctx *
-perf_pt_start_tracer(struct tracer_conf *tr_conf)
+perf_pt_init_tracer(struct tracer_conf *tr_conf)
 {
-    bool failing = false;
     struct tracer_ctx *tr_ctx = NULL;
-    int perf_fd = -1;
-    int clean_sem = 0, clean_thread = 0;
-    struct tracer_ctx *ret = NULL;
+    bool failing = false;
 
     // Allocate and initialise tracer context.
     tr_ctx = malloc(sizeof(*tr_ctx));
     if (tr_ctx == NULL) {
         goto clean;
     }
-    tr_ctx->trace_buf = NULL;
-    tr_ctx->trace_len = 0;
 
-    // Get the perf fd
-    perf_fd = open_perf(tr_conf->target_tid);
-    if (perf_fd == -1) {
+    // Set default values.
+    memset(tr_ctx, 0, sizeof(*tr_ctx));
+    tr_ctx->stop_fds[0] = tr_ctx->stop_fds[1] = -1;
+    tr_ctx->perf_fd = -1;
+
+    // Obtain a file descriptor through which to speak to perf.
+    tr_ctx->perf_fd = open_perf(tr_conf->target_tid);
+    if (tr_ctx->perf_fd == -1) {
+        failing = true;
+        goto clean;
+    }
+
+    // Allocate mmap(2) buffers for speaking to perf.
+    //
+    // These regions are mmap(2)ed from the perf fd in two separate chunks:
+    //
+    //            "Base buffer"
+    // -----------------------------------     ------------------
+    // | header  |       data buffer     | ... |   AUX buffer   |
+    // -----------------------------------     ------------------
+    //           ^ header->data_offset
```

Comment:
Is the AUX buffer immediately after the data buffer? If so, how does one know where the data buffer stops and the AUX buffer starts?

Reply:
They are at consecutive offsets in the file descriptor, but get mmaped to different virtual address ranges (we don't provide an mmap hint).

Reply:
> how does one know where the data buffer stops and the AUX buffer starts?

You hold two different pointers:
```
tr_ctx->base_buf = mmap(...
...
tr_ctx->aux_buf = mmap(...
```

Reply:
OK, I think you need to update the explanation a bit. This isn't very clear at the moment: the figure suggests the two buffers follow each other (somehow).

Reply:
Well, we don't know exactly where there are relative to each other, other than, they are non-overlapping. Anywhere I put the box, there is another place the box could also be :P.

Is it the `...` between the boxes that is confusing?

Reply:
Maybe just don't put the AUX buffer in the diagram then? I don't think it helps.

Reply:
How about a more textual description?

Reply:
Worth a shot.